types/react/index.d.ts

```diff
@@ -306,7 +306,7 @@ declare namespace React {
         // In the future, if we can define its call signature conditionally
         // on the existence of `children` in `P`, then we should remove this.
         readonly props: Readonly<{ children?: ReactNode }> & Readonly<P>;
-        readonly state: null | Readonly<S>;
+        state: null | Readonly<S>;
```

Comment:
I am too late in making this comment, but I don't see why removing this `readonly` is at all a fix. According to TypeScript rules, `readonly` does _not_ affect whether a property is writable _in the constructor_.

Reply:
Also, ideally, it would be nice to be able to use conditional types to make `state` `abstract` if the type of `S` is not empty; this would probably fix all the problems with `state` being left accidentally uninitialized without adding this `null` to the type. The default also should be `null`, and not `{}`, but apparently that breaks some of the mapped types that depend on `S`.

Reply:
You are partially correct. Readonly properties are writable in the constructors of the the class that defines the property. Not in subclasses. Whereas it completely valid and recommended to do so in React. See https://github.com/Microsoft/TypeScript/issues/25134#issuecomment-401120150 for more details.

I would like to emphasize that union with `null` is the correct way to go as react state is `null` unless initialized. Therefore this typescript construction describes it in the best possible way.

Reply:
what @magnetnation said. 

> The readonly should be here to guard against assigning to state outside the constructor, which is invalid according to both TypeScript and React rules.

That wont help, when you'll define state on your class via class property, it will override Base class `readonly state` and will left you just with `state` so you need to provide the readonly explicitly ( this might be issue on TS level )

`readonly state` from base class ( correct errors, but wrong runtime as no state has been defined )
![image](https://user-images.githubusercontent.com/1223799/42089065-1833b156-7b9c-11e8-8599-ad2ea2aa9031.png)

readonly overridden:
![image](https://user-images.githubusercontent.com/1223799/42089139-57bbd876-7b9c-11e8-9013-b61ca293b62a.png)



> The default also should be `null`, and not {}

yeah that's right, also props should be `object` not `{}` etc, which would introduce another breaking changes :)